ddl/column_type_change_test.go

```diff
@@ -1797,3 +1799,59 @@ func (s *testColumnTypeChangeSuite) TestChangeIntToBitWillPanicInBackfillIndexes
 		") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin"))
 	tk.MustQuery("select * from t").Check(testkit.Rows("\x13 1 1.00", "\x11 2 2.00"))
 }
+
+// Close issue #24584
+func (s *testColumnTypeChangeSuite) TestCancelCTCInReorgStateWillCauseGoroutineLeak(c *C) {
```

Comment:
Has this test ever failed before you change `ddl/reorg.go`?

Reply:
because the previous logic won't notify the background worker(notifyReorgCancel)

For the old code, cancel can always succeed, because it doesn't have to notify the worker. (it is still running)
For the test here, we mock an infinite worker here, the cancel succeeds only if the reorg exit.